design-docs/Scalar adapter configuration.md

```diff
@@ -0,0 +1,195 @@
+
+#### Compatibility
+
+We need to keep the current mechanism (`customScalarsMapping` + `addCustomScalarAdapter`) working of course.
+
+Let's make `customScalarsMapping` call `mapScalarToMap` internally and mark it as deprecated.
+
+### Code changes
+
+#### Use the mapping info to generate the code
+
+In `ResponseAdapter` and `VariablesAdapter` code generation, we now have more cases to handle to reference the scalar adapter to use:
+
+```
+if (an Adapter for the scalar is registered with NoArgConstructorAdapterInitializer) {
+    Output the adapter with "()"
+    Note: this is not optimal: they are instantiated each time they are used.
+    Instead we can generate fields in a specific object (e.g. `ScalarAdapters`) and use them in the generated code.
+} else if (an Adapter for the scalar is registered with SingletonAdapterInitializer) {
+    Output the adapter without "()"
+} else if (an Adapter for the scalar is registered with RuntimeAdapterInitializer) {
+    Output the code to lookup the adapter in `customScalarAdapters`
+    Note: if the scalar is built-in, a "Type" class is needed - currently they are not generated.
+} else if (the scalar is a built-in (e.g. `ID`)) {
+    Output the appropriate adapter (same as current behavior)
+} else {
+    Output the code to lookup the adapter in `customScalarAdapters`
+    Note: this is the fallback to current behavior
+}
+```
+
+Generated code will look like this:
+
+```kotlin
+public override fun fromJson(reader: JsonReader, customScalarAdapters: CustomScalarAdapters):
+    GetDateQuery.Data {
+// ...
+
+// Before:
+  0 -> myDate = customScalarAdapters.responseAdapterFor<MyDate>(com.example.type.MyDate.type).nullable().fromJson(reader, customScalarAdapters)
+  1 -> id = StringAdapter.fromJson(reader, customScalarAdapters)
+  2 -> myLong = customScalarAdapters.responseAdapterFor<Long>(MyLong.type).nullable().fromJson(reader, customScalarAdapters)
+
+// After:
+  0 -> myDate = com.example.MyDateAdapter().nullable().fromJson(reader, customScalarAdapters)
+  1 -> id = com.apollographql.apollo3.api.LongAdapter.fromJson(reader, customScalarAdapters)
+  2 -> myLong = StringAdapter.fromJson(reader, customScalarAdapters)
+  
+// ...
+```
+
+#### Update and store mapping information
+
+Currently, `customScalarsMapping` in `IrBuilder` (and `Options`) is a `Map<String, String>`.
+Let's change it to a `Map<String, ScalarInfo>`, with `ScalarInfo` being `data class ScalarInfo(val targetName: String, val adapterInitializer: AdapterInitializer)`.
+
+We need to pass it to `KotlinResolver` / `JavaResolver` because this is where the logic to get the adapter lies (`adapterInitializer()`, `nonNullableAdapterInitializer()`). 
+
+This can be done by passing it to `KotlinCodeGen` / `JavaCodeGen` which instantiate the resolvers.
+
+#### Remaining points
+
+- Avoid instantiating the no arg constructors adapters at each use
+  -> Generate fields in a specific object (e.g. `ScalarAdapters`) and use them in the generated code?
```

Comment:
We've historically tried to avoid that kind of lazy logic because it requires synchronization for proper multi-threading support. I think making a new instance everytime is ok. The caller can always use a singleton if they want to optimize

Reply:
Oh ok - I actually have a prototype that works, but you're right about potential risk / it being not that useful. Removing (will make the implementation simpler) 👍